### ISSUE TITLE
Bug: useEffect Timing changes depending on if Portal was rendered

### ISSUE THREAD

**Issue by jquense** [facebook/react#20074]
This is a weird one. Basically, if you add an event listener to the document in an effect that was triggered by an event. e.g. `click` toggles some state, which triggers an effect, which adds a `click` handler to the document. In the normal case the new event handler will "miss" the triggering event, e.g. the added click handler won't respond to the click event that triggered it being added (omg). 

HOWEVER, if you render a portal first, the timing changes slightly and the added event handler will see the current event.

React version: 17

## Steps To Reproduce

https://codesandbox.io/s/react-playground-forked-cyt0f?file=/index.js

1. Click the "show Message" button to see a message toggle in and out
2. Click the "Render Portal" button (see a portal rendered into the body)
3. Click the "show Message" button again and notice nothing happens

The reason for the final behavior is the click event both opens and closes the message, (calls set state twice)

Link to code example:

https://codesandbox.io/s/react-playground-forked-cyt0f?file=/index.js

## The expected behavior

That they be consistent

**Comment by gaearon** [comment]
17 only?

**Comment by jquense** [comment]
17 only! downstream issue  https://github.com/react-bootstrap/react-bootstrap/issues/5409

**Comment by jquense** [comment]
well tbh maybe this was present but not observable since before 17 you couldn't attach an event listener "higher up" so there wasn't any way to get in front of the current event

**Comment by gaearon** [comment]
What sounds like a bug here is that `useEffect` is called synchronously at all. It should normally be deferred — which would avoid this issue. So it sounds like there's an unexpected effect flush. Would be nice to dig into why.

**Comment by gaearon** [comment]
I'm sooo confused. I can't reproduce this anymore. The only thing I did was to update Chrome...

**Comment by gaearon** [comment]
Can you still repro it?

**Comment by jquense** [comment]
bah sorry @gaearon i was goofing around trying to find a workaround and didn't realize it was saving instead of forking. It should be back to being reproducible

**Comment by gaearon** [comment]
Ok at least I can still sleep at night

**Comment by gaearon** [comment]
What happens here is that the native event bubbles:

- On the root container
- On the body (portal target)

This is why we get into dispatchEvent twice. Root container pass schedules an effect, and body pass flushes it.

Seems like a bug.

**Comment by gaearon** [comment]
To follow up on this — it does seem like a bug but we can't commit to prioritizing it at the moment. There will be a bunch of work done to clean up and simplify related parts of the code in the coming months so we'll likely address it then. Have you found any userland workaround, or is this a hard blocker for you?

**Comment by jquense** [comment]
We've got a workaround that's simple and effective in theory [here](https://github.com/react-bootstrap/react-overlays/blob/master/src/useRootClose.ts#L91). Need to let folks test it out a bit though. Alternatively I think adding listeners in a` setImmediate` would work albeit pretty annoying to manage in an effect.

For anyone looking out current workaround is 
```js
useEffect(() => {
  // Note this depends on a deprecated but extremely well 
  // supported quirk of the web platform, this may be less effective 
  // for other event types, YMMV
  const currentEvent = window.event

  element.addEventListener('click', (e) => {
     if (e === currentEvent) {
        currentEvent = null;
        return
     } 
     // do the work
  })

  // ...clean up
})
```

**Comment by karesztrk** [comment]
I made it working by using the good old [stopPropagation](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation). Attach an `onClick` listener to the dropdown and stop the event propagating.

**Comment by eps1lon** [comment]
What I noticed is that for an `act(() => trigger.click())` everything works fine but an actual user click causes the issue described: https://codesandbox.io/s/material-ui-issue-forked-gjmkc?file=/src/index.js

**Comment by eps1lon** [comment]
@gaearon It seems to me that the assertions in https://github.com/facebook/react/blob/c59c3dfe554dafb64864f3bbcfff6ffe51f32275/packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js#L1705-L1722 are not correct.

The update that is scheduled in the passive effect is deferred as intended when flushing sync but the effect itself should not run when flushing synchronously.

IMO 
```diff
diff --git a/packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js b/packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
index 79794bbea7..6ab75eac6d 100644
--- a/packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
+++ b/packages/react-reconciler/src/__tests__/ReactHooksWithNoopRenderer-test.js
@@ -1706,14 +1706,16 @@ describe('ReactHooksWithNoopRenderer', () => {
         ReactNoop.flushSync(() => {
           _updateCount(2);
         });
+
+        expect(Scheduler).toHaveYielded(['Count: 2']);
       });
 
       // As a result we, somewhat surprisingly, commit them in the opposite order.
       // This should be fine because any non-discrete set of work doesn't guarantee order
       // and easily could've happened slightly later too.
       expect(Scheduler).toHaveYielded([
-        'Will set count to 1',
         'Count: 2',
+        'Will set count to 1',
         'Count: 1',
       ]);
 ```
would make more sense here. 

The test comment acknowledges that this is the opposite order and it seems to me that this issue is a good example why this order is problematic: Any side-effect that touches the  DOM directly is out-of-order.

**Comment by mickaelzhang** [comment]
@jquense Not sure about the intended behaviour but it seems like having the following solve your issue (at least temporary):
```js
document.addEventListener("click", clickHandler, { capture: true })
```

On step 3, it displays the message correctly

**Comment by WesCossick** [comment]
You can also fix the issue by wrapping `document.addEventListener()` in a `setTimeout()`, like so:

```ts
setTimeout(() => {
    document.addEventListener('click', clickHandler);
}, 1);
```

**Comment by BrinsterAman** [comment]
Is there a fix in progress, for upcoming versions, for this bug? I am facing this issue https://github.com/mui-org/material-ui/issues/24783 , which cant be resolved because of this.

**Comment by eps1lon** [comment]
> I am facing this issue mui-org/material-ui#24783 , which cant be resolved because of this.

Could you explain in the Material-UI issue why this can't be resolved for you? We were under the impression that we worked around this bug for our specific usage.

For updates on a fix, https://github.com/facebook/react/issues/20074#issuecomment-717318697 is the latest update. From what I gather the team is already experimenting with potential fixes. You can subscribe to this (https://github.com/facebook/react/issues/20074) github issue to get notified about updates. Asking for updates does not speed up delivery of said update.

**Comment by gaearon** [comment]
The team has taken a closer look at this now that the related refactors are landing.

We can't fix it in a general case without relying on some heuristic. Making all effects synchronous would be bad. Making all of them asynchronous also doesn't work because sometimes we need to flush earlier. The need to flush earlier relates to handling discrete events (like click) that need to be able to observe the already flushed state. 

A reasonable compromise is to make them synchronous at the end of discrete events (such as click), but asynchronous in other cases. This mirrors the recent changes we've made to state updates in Concurrent Mode to make it easier to adopt. So we added this heuristic in Concurrent Mode: https://github.com/facebook/react/pull/21150. I checked that this issue [doesn't repro](https://codesandbox.io/s/react-playground-forked-mp0jo?file=/package.json) in Concurrent Mode with the latest master commit and now consistently does _not_ show the message (because clicks lead to synchronous effects).

We've considered backporting this behavior to React 17. But at this point it's very risky because it's such a subtle change. It's also risky because it would be the first time that we rely on `window.event.type`-based heuristics outside of Concurrent Mode. So it's kind of a new territory. Given that it's easy to work around, it doesn't seem like this risk is warranted in React 17.

So we've concluded that, while unfortunate, it is safer to not try to fix it in React 17. But Concurrent Mode will have a more predictable behavior here. (Note that it's also quite likely that React 16 _also_ had cases with over-flushing, but triggered under different conditions.) I know this isn't the resolution that most people were hoping for.
